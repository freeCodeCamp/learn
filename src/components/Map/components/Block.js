import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import Link, { navigateTo } from 'gatsby-link';

import ga from '../../../analytics';
import { makeExpandedBlockSelector, toggleBlock } from '../redux';
import { toggleMapModal } from '../../../redux/app';
import Caret from '../../icons/Caret';

const mapStateToProps = (state, ownProps) => {
  const expandedSelector = makeExpandedBlockSelector(ownProps.blockDashedName);

  return createSelector(expandedSelector, isExpanded => ({ isExpanded }))(
    state
  );
};

const mapDispatchToProps = dispatch =>
  bindActionCreators({ toggleBlock, toggleMapModal }, dispatch);

const propTypes = {
  blockDashedName: PropTypes.string,
  challenges: PropTypes.array,
  intro: PropTypes.shape({
    fields: PropTypes.shape({ slug: PropTypes.string.isRequired }),
    frontmatter: PropTypes.shape({
      title: PropTypes.string.isRequired,
      block: PropTypes.string.isRequired
    })
  }),
  isExpanded: PropTypes.bool,
  toggleBlock: PropTypes.func.isRequired,
  toggleMapModal: PropTypes.func.isRequired
};

export class Block extends PureComponent {
  constructor(...props) {
    super(...props);

    this.handleBlockClick = this.handleBlockClick.bind(this);
    this.handleChallengeClick = this.handleChallengeClick.bind(this);
    this.renderChallenges = this.renderChallenges.bind(this);
  }

  handleBlockClick() {
    const { blockDashedName, challenges, toggleBlock } = this.props;
    const blockPath = challenges[0].fields.slug
      .split('/')
      .slice(0, -1)
      .join('/');
    toggleBlock(blockDashedName);
    ga.event({
      category: 'Map Block Click',
      action: blockDashedName
    });
    return navigateTo(blockPath);
  }

  handleChallengeClick(slug) {
    return () => {
      this.props.toggleMapModal();
      return ga.event({
        category: 'Map Challenge Click',
        action: slug
      });
    };
  }

  renderChallenges(intro, challenges) {
    // TODO: Split this into a Challenge Component and add tests
    return [intro].concat(challenges).map(challenge => (
      <li
        className='map-challenge-title'
        key={'map-challenge' + challenge.fields.slug}
        >
        <Link
          onClick={this.handleChallengeClick(challenge.fields.slug)}
          to={challenge.fields.slug}
          >
          {challenge.title || challenge.frontmatter.title}
        </Link>
      </li>
    ));
  }

  render() {
    const { challenges, isExpanded, intro } = this.props;
    const { blockName } = challenges[0].fields;
    return (
      <li className={`block ${isExpanded ? 'open' : ''}`}>
        <div className='map-title' onClick={this.handleBlockClick}>
          <Caret />
          <h5>{blockName}</h5>
        </div>
        <ul>{isExpanded ? this.renderChallenges(intro, challenges) : null}</ul>
      </li>
    );
  }
}

Block.displayName = 'Block';
Block.propTypes = propTypes;

export default connect(mapStateToProps, mapDispatchToProps)(Block);
