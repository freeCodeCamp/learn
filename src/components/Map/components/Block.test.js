/* global expect */

import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';

import { Block } from './Block';
import mockNodes from '../../../__mocks__/map-nodes';
import mockIntroNodes from '../../../__mocks__/intro-nodes';
import mockCompleted from '../../../__mocks__/completedChallengesMock';

Enzyme.configure({ adapter: new Adapter() });
const renderer = new ShallowRenderer();

test('<Block /> not expanded snapshot', () => {
  const toggleSpy = sinon.spy();
  const toggleMapSpy = sinon.spy();
  const componentToRender = (
    <Block
      blockDashedName='block-a'
      challenges={mockNodes.filter(node => node.block === 'block-a')}
      completedChallenges={mockCompleted}
      intro={mockIntroNodes[0]}
      isExpanded={false}
      toggleBlock={toggleSpy}
      toggleMapModal={toggleMapSpy}
    />
  );
  const component = renderer.render(componentToRender);

  expect(component).toMatchSnapshot('block-not-expanded');
});

test('<Block expanded snapshot', () => {
  const toggleSpy = sinon.spy();
  const toggleMapSpy = sinon.spy();
  const componentToRender = (
    <Block
      blockDashedName='block-a'
      challenges={mockNodes.filter(node => node.block === 'block-a')}
      completedChallenges={mockCompleted}
      intro={mockIntroNodes[0]}
      isExpanded={true}
      toggleBlock={toggleSpy}
      toggleMapModal={toggleMapSpy}
    />
  );

  const component = renderer.render(componentToRender);

  expect(component).toMatchSnapshot('block-expanded');
});

test('<Block />  should handle toggle clicks correctly', () => {
  const toggleSpy = sinon.spy();
  const toggleMapSpy = sinon.spy();
  const props = {
    blockDashedName: 'block-a',
    challenges: mockNodes.filter(node => node.block === 'block-a'),
    completedChallenges: mockCompleted,
    intro: mockIntroNodes[0],
    isExpanded: false,
    toggleBlock: toggleSpy,
    toggleMapModal: toggleMapSpy
  };

  const componentToRender = <Block {...props} />;

  const enzymeWrapper = Enzyme.shallow(componentToRender);

  expect(toggleSpy.called).toBe(false);
  expect(
    enzymeWrapper
      .find('.map-title')
      .find('h5')
      .text()
  ).toBe('Block A');

  enzymeWrapper.find('.map-title').simulate('click');

  expect(toggleSpy.called).toBe(true);
  expect(toggleSpy.calledWithExactly('block-a')).toBe(true);

  enzymeWrapper.setProps({ ...props, isExpanded: true });

  expect(
    enzymeWrapper
      .find('.map-title')
      .find('h5')
      .text()
  ).toBe('Block A');
  expect(enzymeWrapper.find('ul').length).toBe(1);
});
