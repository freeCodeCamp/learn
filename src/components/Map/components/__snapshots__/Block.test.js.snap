// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Block /> not expanded snapshot: block-not-expanded 1`] = `
<li
  className="block "
>
  <div
    className="map-title"
    onClick={[Function]}
  >
    <Caret />
    <h5>
      Block A
    </h5>
  </div>
  <ul />
</li>
`;

exports[`<Block expanded snapshot: block-expanded 1`] = `
<li
  className="block open"
>
  <div
    className="map-title"
    onClick={[Function]}
  >
    <Caret />
    <h5>
      Block A
    </h5>
  </div>
  <ul>
    <li
      className="map-challenge-title"
    >
      <span
        className="badge map-badge"
      />
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a"
      >
        Introduction to Block A
      </Unknown>
    </li>
    <li
      className="map-challenge-title map-challenge-title-completed"
    >
      <span
        className="badge map-badge"
      >
        <GreenPass
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a/challenge-one"
      >
        Challenge One
      </Unknown>
    </li>
    <li
      className="map-challenge-title"
    >
      <span
        className="badge map-badge"
      >
        <GreenNotCompleted
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a/challenge-two"
      >
        Challenge Two
      </Unknown>
    </li>
    <li
      className="map-challenge-title"
    >
      <span
        className="badge map-badge"
      >
        <GreenNotCompleted
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a/challenge-one"
      >
        Challenge One
      </Unknown>
    </li>
    <li
      className="map-challenge-title map-challenge-title-completed"
    >
      <span
        className="badge map-badge"
      >
        <GreenPass
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a/challenge-two"
      >
        Challenge Two
      </Unknown>
    </li>
  </ul>
</li>
`;
